src/NuGet.Clients/NuGet.SolutionRestoreManager/ErrorListTableDataSource.cs

```diff
@@ -0,0 +1,205 @@
+        }
+
+        /// <summary>
+        /// Clear only nuget entries.
+        /// </summary>
+        public void ClearNuGetEntries()
+        {
+            EnsureInitialized();
+
+            var subscriptions = _subscriptions;
+            foreach (var subscription in subscriptions)
+            {
+                subscription.RunWithLock((sink) =>
+                {
+                    var entries = _errorList.TableControl.Entries.Where(IsNuGetEntry).ToArray();
+
+                    sink.RemoveEntries(entries);
+                });
+            }
+        }
+
+        /// <summary>
+        /// Add error list entries
+        /// </summary>
+        public void AddEntries(params ErrorListTableEntry[] entries)
+        {
+            if (entries.Length > 0)
+            {
+                EnsureInitialized();
+
+                var subscriptions = _subscriptions;
+                foreach (var subscription in subscriptions)
+                {
+                    subscription.RunWithLock((sink) =>
+                    {
+                        sink.AddEntries(entries.ToList(), removeAllEntries: false);
+                    });
+                }
+            }
+        }
+
+        /// <summary>
+        /// Show error window.
+        /// </summary>
+        public void BringToFront()
+        {
+            EnsureInitialized();
+
+            // Give the error list focus.
+            var vsErrorList = _errorList as IVsErrorList;
+            vsErrorList?.BringToFront();
+        }
+
+        // Lock before calling
+        private void EnsureInitialized()
+        {
+            if (!_initialized)
+            {
+                // Double check around locking since this is called often.
+                lock (_initLockObj)
+                {
+                    if (!_initialized)
+                    {
+                        NuGetUIThreadHelper.JoinableTaskFactory.Run(async () =>
+                        {
+                            await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+                            // Get the error list service from the UI thread
+                            _errorList = _serviceProvider.GetService(typeof(SVsErrorList)) as IErrorList;
+                            _tableManager = _errorList.TableControl.Manager;
+
+                            _tableManager.AddSource(this);
+                            _initialized = true;
+                        });
+                    }
+                }
+            }
+        }
+
+        private static bool IsNuGetEntry(ITableEntryHandle entry)
+        {
+            object sourceObj;
+            return (entry != null
+                && entry.TryGetValue(StandardTableColumnDefinitions.ErrorSource, out sourceObj)
+                && StringComparer.Ordinal.Equals(ErrorListTableEntry.ErrorSouce, (sourceObj as string)));
+        }
+
+        public void Dispose()
+        {
+            if (_initialized)
```

Comment:
This needs to be placed under the lock. Otherwise there's a race condition window. Also have to reset `_initialized`.